opendevin/events/stream.py

```diff
@@ -64,15 +66,17 @@ def get_event(self, id: int) -> Event:
 
     def subscribe(self, id: EventStreamSubscriber, callback: Callable):
         if id in self._subscribers:
-            raise ValueError('Subscriber already exists: ' + id)
+            self._subscribers[id].append(callback)
         else:
-            self._subscribers[id] = callback
+            self._subscribers[id] = [callback]
```

Comment:
I've gone back and forth on this, but I think we should raise an exception here. It's indicative of something going very wrong. Not raising here caused me to miss a pretty severe bug in another PR

Reply:
It’s by design when you have (multi-layer) delegates - so I maintain a stack here.

Reply:
What if the parent AgentController just calls `on_event` to its child? Then we only subscribe once from the top-level AgentController.

We should probably pass None as the event_stream to the Delegate in order to prevent issues

Reply:
> What if the parent AgentController just calls on_event to its child? Then we only subscribe once from the top-level AgentController.

That won't work because AgentController needs to change its state in the callback.

Reply:
> What if the parent AgentController just calls on_event to its child?

That's a viable approach I think, but it would be hard to implement/manage multi-layer agent delegate. If grandpa delegates to parent, which in turn delegates to a child, e.g. main agent -> database agent -> postgres migration agent, it would be hard/inelegant to let the root AgentController handle everything.

Reply:
What if we create a new EventStream for every delegated agent with the sid like `id + '-delegate-xxagent'`? 
If we just pass None as the event_stream to the Delegate like Robert said, I'm not sure whether that means delegated agents are not managed and driven by events?

Reply:
> What if we create a new EventStream for every delegated agent

That's another viable approach! However, I do feel like EventStream should be session-based, not agent based.

Reply:
> EventStream should be session-based, not agent based.

After rethink it,  and with the info in this pr(https://github.com/OpenDevin/OpenDevin/pull/1810)，
I draw a pic to help me understand the whole thing: 
![image](https://github.com/OpenDevin/OpenDevin/assets/1708914/712b4e80-d07d-4dc1-8770-78a5ce7dcf95)

I agree with you. Delegate shares event_stream/runtime with the main controller. 

I'm not sure whethre add a new enum like `DELEGATE_AGENT_CONTROLLER` in `EventStreamSubscriber` is an effective way to solve.

Reply:
> I'm not sure whethre add a new enum like DELEGATE_AGENT_CONTROLLER in EventStreamSubscriber is an effective way to solve.

I don't think this works unfortunately, because you could have any number of delegates...

Reply:
> That's a viable approach I think, but it would be hard to implement/manage multi-layer agent delegate. If grandpa delegates to parent, which in turn delegates to a child, e.g. main agent -> database agent -> postgres migration agent, it would be hard/inelegant to let the root AgentController handle everything.

Can you explain more the issue here? Each AgentController would get to see every event, so it should be flexible enough to handle every possible case.

Reply:
To be clear, the logic I'm envisioning is something like
```python
def on_event(evt):
  if instanceof(evt, Foo):
    # ...all the normal logic...
  if self.delegate:
    self.delegate.on_event(evt)
```

Reply:
Oh that’s a great idea! Will try

Reply:
Done. I am not happy with the current implementation either (need to check whether itself is a delegate before it subscribe/unsubscribe) but looks like there's no perfect approach!

Reply:
I changed it back to the original implementation - let each AgentController worry about their own subscriptions and listeners. I added an `append` parameter to `event_stream.subscribe` method, and only delegates could set this parameter to true. This should largely prevent bugs like duplicate subscription...